cohortextractor/query_engines/mssql.py

```diff
@@ -49,10 +50,65 @@ def get_primary_table(query):
     return get_joined_tables(query)[0]
 
 
+# MS-SQL can misinterpret ISO dates, depending on its localisation settings so
+# we need to use particular date formats which we know will be consistently
+# interpreted. We do this by defining custom SQLAlchemy types. See:
+# https://github.com/opensafely-core/cohort-extractor-v2/issues/92
+# http://msdn.microsoft.com/en-us/library/ms180878.aspx
+# https://stackoverflow.com/a/25548626/559140
+class _MSSQLDateTimeBase:
+    text_type = sqlalchemy.types.Text()
+
+    def process_bind_param(self, value, dialect):
+        """
+        Convert a Python value to a form suitable for passing as a parameter to
+        the database connector
+        """
+        if value is None:
+            return None
+        # We accept ISO formated strings as well
+        if isinstance(value, str):
+            value = self.date_type.fromisoformat(value)
+        if not isinstance(value, self.date_type):
+            raise TypeError(f"Expected {self.date_type} or str got: {value!r}")
+        return value.strftime(self.format_str)
+
+    def process_literal_param(self, value, dialect):
+        """
+        Convert a Python value into an escaped string suitable for
+        interpolating directly into an SQL string
+        """
+        value = self.process_bind_param(value, dialect)
+        # Use the Text literal processor to handle quoting and esacping
+        literal_processor = self.text_type.literal_processor(dialect)
+        return literal_processor(value)
```

Comment:
Why don't we have to do any conversion or checking here?

Reply:
That all happens in this line which calls the method defined above:
```
value = self.process_bind_param(value, dialect)
```
So we're doing all the things we normally do to prepare a parameter to be passed to `cursor.execute()` and then we're invoking the Text type to convert the resulting string into something that can be safely interpolated.

I agree I could definitely comment this better!

Reply:
Oh, I just read through that first line.